beacon_node/network/src/subnet_service/tests/mod.rs

```diff
@@ -0,0 +1,568 @@
+                .build()
+                .expect("should build"),
+        );
+        Self { chain }
+    }
+}
+
+pub fn recent_genesis_time() -> u64 {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .as_secs()
+}
+
+fn get_logger() -> Logger {
+    NullLoggerBuilder.build().expect("logger should build")
+}
+
+lazy_static! {
+    static ref CHAIN: TestBeaconChain = TestBeaconChain::new_with_system_clock();
+}
+
+fn get_attestation_service() -> AttestationService<TestBeaconChainType> {
+    let log = get_logger();
+    let config = NetworkConfig::default();
+
+    let beacon_chain = CHAIN.chain.clone();
+
+    AttestationService::new(beacon_chain, &config, &log)
+}
+
+fn get_sync_committee_service() -> SyncCommitteeService<TestBeaconChainType> {
+    let log = get_logger();
+    let config = NetworkConfig::default();
+
+    let beacon_chain = CHAIN.chain.clone();
+
+    SyncCommitteeService::new(beacon_chain, &config, &log)
+}
+
+// gets a number of events from the subscription service, or returns none if it times out after a number
+// of slots
+async fn get_events<S: Stream<Item = SubnetServiceMessage> + Unpin>(
+    stream: &mut S,
+    num_events: Option<usize>,
+    num_slots_before_timeout: u32,
+) -> Vec<SubnetServiceMessage> {
+    let mut events = Vec::new();
+
+    let collect_stream_fut = async {
+        loop {
+            if let Some(result) = stream.next().await {
+                events.push(result);
+                if let Some(num) = num_events {
+                    if events.len() == num {
+                        return;
+                    }
+                }
+            }
+        }
+    };
+
+    tokio::select! {
+        _ = collect_stream_fut => events,
+        _ = tokio::time::sleep(
+        Duration::from_millis(SLOT_DURATION_MILLIS) * num_slots_before_timeout,
+    ) => events
+    }
+}
+
+mod attestation_service {
+    use super::*;
+
+    fn get_subscription(
+        validator_index: u64,
+        attestation_committee_index: CommitteeIndex,
+        slot: Slot,
+        committee_count_at_slot: u64,
+    ) -> ValidatorSubscription {
+        let is_aggregator = true;
+        ValidatorSubscription {
+            validator_index,
+            attestation_committee_index,
+            slot,
+            committee_count_at_slot,
+            is_aggregator,
+        }
+    }
+
+    fn get_subscriptions(
+        validator_count: u64,
+        slot: Slot,
+        committee_count_at_slot: u64,
+    ) -> Vec<ValidatorSubscription> {
+        (0..validator_count)
+            .map(|validator_index| {
+                get_subscription(
+                    validator_index,
+                    validator_index,
+                    slot,
+                    committee_count_at_slot,
+                )
+            })
+            .collect()
+    }
+
+    #[tokio::test]
+    async fn subscribe_current_slot_wait_for_unsubscribe() {
```

Comment:
I got a test failure here on `altair-vc` that looks kind of non-deterministic:

https://github.com/sigp/lighthouse/runs/3187550970

Reply:
Seems to be fixed now :+1: